### PR TITLE
improvement: avoid ub and delele unnecessary unsafe

### PR DESCRIPTION
**Pull request by xudong963**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

**1. Avoid undefine behavior**: 
In some cases (only group by or `select distinct`), in transform aggregation, `aggregate_functions` are empty, so in `get_layout_offsets`, the align of the generated `Layout` is 0, which will cause UB when allocating memory.

**2. Delete unnecessary unsafe**:
Replace `from_size_align_unchecked` with `from_size_align` and delete unsafe.

## Changelog

- Improvement

## Related Issues

Fixes #issue

### CODE REVIEW

File: query/src/pipelines/new/processors/transforms/aggregator/aggregator_final.rs [hunk 1]
```diff
@@ -69,23 +70,29 @@ pub struct FinalAggregator<
...
 {
-    pub fn create(method: Method, params: Arc<AggregatorParams>, ctx: Arc<QueryContext>) -> Self {
+    pub fn create(
+        method: Method,
+        params: Arc<AggregatorParams>,
+        ctx: Arc<QueryContext>,
+    ) -> Result<Self> {
         let state = method.aggregate_state();
         let temp_place = if params.aggregate_functions.is_empty() {
             0.into()
```
Comment (sundy-li) [inline query/src/pipelines/new/processors/transforms/aggregator/aggregator_final.rs:80]: Could we make temp_place as Option ?

File: query/src/pipelines/new/processors/transforms/aggregator/aggregator_final.rs [hunk 2]
```diff
@@ -105,9 +112,10 @@ impl<Method: HashMethod + PolymorphicKeysHelper<Method> + Send> FinalAggregator<
 
             match inserted {
                 true => {
-                    let place = state.alloc_layout2(params);
-                    places.push(place);
-                    entity.set_state_value(place.addr());
+                    if let Some(place) = state.alloc_layout2(params) {
```
Comment (zhang2014) [inline query/src/pipelines/new/processors/transforms/aggregator/aggregator_final.rs:115]: What happen with layout is None?

File: query/src/pipelines/new/processors/transforms/aggregator/aggregator_single_key.rs [hunk 3]
```diff
@@ -54,26 +54,31 @@ pub struct SingleStateAggregator<const FINAL: bool> {
...
-                })
-                .collect()
-        };
-
-        let places = get_places();
-        let temp_places = get_places();
 
+        let mut places: Vec<StateAddr> = vec![];
+        let mut temp_places: Vec<StateAddr> = vec![];
+        if !params.offsets_aggregate_states.is_empty() {
```
Comment (zhang2014) [inline query/src/pipelines/new/processors/transforms/aggregator/aggregator_single_key.rs:60]: replace with `assert!`

**Comment by zhang2014** [comment]
Need performance test for this pr.

**Comment by zhang2014** [comment]
```
In some cases (only group by or select distinct), in transform aggregation, aggregate_functions are empty, so in get_layout_offsets, the align of the generated Layout is 0, which will cause UB when allocating memory.
```
maybe this should not happen. When there is no aggregate function, we will not memory allocation.

**Comment by zhang2014** [comment]
Performance test is ok on macos:

before:

```
mysql> SELECT max(number) FROM numbers_mt(10000000000) GROUP BY number % 3;
+-------------+
| max(number) |
+-------------+
|  9999999999 |
|  9999999997 |
|  9999999998 |
+-------------+
3 rows in set (5.89 sec)
Read 10000000000 rows, 74.51 GiB in 5.886 sec., 1.7 billion rows/sec., 12.66 GiB/sec.

mysql> SELECT max(number), sum(number) FROM numbers_mt(10000000000) GROUP BY number % 3;
+-------------+----------------------+
| max(number) | sum(number)          |
+-------------+----------------------+
|  9999999999 | 16666666668333333333 |
|  9999999997 | 16666666661666666667 |
|  9999999998 | 16666666665000000000 |
+-------------+----------------------+
3 rows in set (7.49 sec)
Read 10000000000 rows, 74.51 GiB in 7.489 sec., 1.34 billion rows/sec., 9.95 GiB/sec.

mysql> SELECT max(number), sum(number) FROM numbers_mt(10000000000) GROUP BY number % 3, number % 4, number % 5 LIMIT 10;
+-------------+--------------------+
| max(number) | sum(number)        |
+-------------+--------------------+
|  9999999960 | 833333331666666660 |
|  9999999941 | 833333328500000006 |
|  9999999984 | 833333335666666668 |
|  9999999963 | 833333332166666661 |
|  9999999962 | 833333331999999994 |
|  9999999954 | 833333330666666664 |
|  9999999942 | 833333328666666672 |
|  9999999989 | 833333336500000003 |
|  9999999990 | 833333336666666670 |
|  9999999994 | 833333337333333338 |
+-------------+--------------------+
10 rows in set (9.88 sec)
Read 10000000000 rows, 74.51 GiB in 9.861 sec., 1.01 billion rows/sec., 7.56 GiB/sec.
```

after:
```
mysql> SELECT max(number) FROM numbers_mt(10000000000) GROUP BY number % 3;
+-------------+
| max(number) |
+-------------+
|  9999999999 |
|  9999999997 |
|  9999999998 |
+-------------+
3 rows in set (5.91 sec)
Read 10000000000 rows, 74.51 GiB in 5.899 sec., 1.7 billion rows/sec., 12.63 GiB/sec.


mysql> SELECT max(number), sum(number) FROM numbers_mt(10000000000) GROUP BY number % 3;
+-------------+----------------------+
| max(number) | sum(number)          |
+-------------+----------------------+
|  9999999999 | 16666666668333333333 |
|  9999999997 | 16666666661666666667 |
|  9999999998 | 16666666665000000000 |
+-------------+----------------------+
3 rows in set (7.53 sec)
Read 10000000000 rows, 74.51 GiB in 7.514 sec., 1.33 billion rows/sec., 9.92 GiB/sec.

mysql> SELECT max(number), sum(number) FROM numbers_mt(10000000000) GROUP BY number % 3, number % 4, number % 5 LIMIT 10;
+-------------+--------------------+
| max(number) | sum(number)        |
+-------------+--------------------+
|  9999999960 | 833333331666666660 |
|  9999999941 | 833333328500000006 |
|  9999999984 | 833333335666666668 |
|  9999999963 | 833333332166666661 |
|  9999999962 | 833333331999999994 |
|  9999999954 | 833333330666666664 |
|  9999999942 | 833333328666666672 |
|  9999999989 | 833333336500000003 |
|  9999999990 | 833333336666666670 |
|  9999999994 | 833333337333333338 |
+-------------+--------------------+
10 rows in set (9.97 sec)
Read 10000000000 rows, 74.51 GiB in 9.961 sec., 1 billion rows/sec., 7.48 GiB/sec.
```

**Comment by xudong963** [comment]
> maybe this should not happen. When there is no aggregate function, we will not memory allocation.

Seems `Bump` will still allocate layout, then call `layout.align()`, if align is 0, UB will happen.

File: query/src/pipelines/new/processors/transforms/aggregator/aggregator_final.rs (hunk 2 above)
Comment (xudong963) [reply query/src/pipelines/new/processors/transforms/aggregator/aggregator_final.rs:115]: It means there is no state and no memory address will be returned.

File: query/src/pipelines/new/processors/transforms/aggregator/aggregator_single_key.rs (hunk 3 above)
Comment (xudong963) [reply query/src/pipelines/new/processors/transforms/aggregator/aggregator_single_key.rs:60]: Don't quite understand why `assert!` need be used here

**Review by sundy-li — APPROVED** [review]

**Review by zhang2014 — APPROVED** [review]

File: query/src/pipelines/new/processors/transforms/aggregator/aggregator_single_key.rs (hunk 3 above)
Comment (zhang2014) [reply query/src/pipelines/new/processors/transforms/aggregator/aggregator_single_key.rs:60]: When the aggregate function is empty, we will not execute this part. So in this case, this is a bug.

Comment (xudong963) [reply query/src/pipelines/new/processors/transforms/aggregator/aggregator_single_key.rs:60]: Originally, only when the group by is empty will go here, if the aggregation function is also empty, then certainly will not go here, if it comes is a bug. Thanks for your explanation!

**Comment by xudong963** [comment]
@mergify update

**Pull request merged**